src/base.js

```diff
@@ -628,9 +631,179 @@ class Base {
+      }
+      if (this.isTaggedMatrixMessage(messageText) || isFilenameTagged(attachedFilePath || '')) {
+        info('it is from matrix, so just ignore it.');
+        isThirdParty = false;
+      }
+      info('it is from 3rd party client');
+    }
+    return isThirdParty;
+  }
+
+  async videoDimensions(videoFile) {
+
+    let getMetadata = (file) => new Promise((resolve, reject) => {
+      ffmpeg.ffprobe(file, function(err, metadata) {
+        if (err) { reject(err); }
+        else { resolve(metadata); }
+      });
+    });
+
+    try {
+      let metadata = await getMetadata(videoFile);
+
+      // video stream isn't necessarily the first one. loop through the streams
+      // and look for codec_type: 'video'
+      let videoStream;
+      for (let i in metadata.streams) {
+        let stream = metadata.streams[i];
+        if (stream.codec_type === "video") {
+          videoStream = stream;
+          break;
+        }
+      }
+
+      if (videoStream) {
+        var w = videoStream.width;
+        var h = videoStream.height;
+
+        if ("rotation" in videoStream) {
+          let r = videoStream.rotation;
+
+          // Not actually sure what the possible rotation values are. Hopefully this covers it.
+          if (r === "0" || r === "-0" || r === "180" || r === "-180") { var isRotated = false; }
+          else { var isRotated = true; }
+        }
+      }
+
+      return (isRotated ? { w: h, h: w } : { w: w, h: h });
+    } catch {
+      return {w: 0, h: 0};
+    }
+  }
+
+  // Payload can include a url, path, or buffer. Mimetype is optional 
+  // (we'll attempt to figure it out unless it's set explicitly), but it's best to set it 
+  // when sending a buffer if possible. If the mimetype isn't set and we can't figure it out
+  // the attachement will be sent as an m.file message.
+  async handleThirdPartyRoomMessageWithAttachment(payload) {
+    const { info, warn } = debug(this.handleThirdPartyRoomMessageWithAttachment.name);
+    info('handling third party room message with attachment', payload);
+    let {
+      roomId,
+      senderName,
+      senderId,
+      avatarUrl,
+      text,
+      url, path, buffer,
+      mimetype,
+    } = payload;
+
+    const matrixRoomId = await this.getOrCreateMatrixRoomFromThirdPartyRoomId(roomId);
+    const client = await this.getUserClient(matrixRoomId, senderId, senderName, avatarUrl);
+
+    if (!this.messageIsFromThirdParty(senderId, text, url || path)) {
+      return;
+    }
+
+    if (!mimetype) mimetype = mime.lookup(url || path);
+
+    let upload = async(buffer, opts) => {
+      const res = await client.uploadContent(buffer, Object.assign({
+        name: text,
+        type: mimetype,
+        rawResponse: false
+      }, opts || {}));
+      return {
+        content_uri: res.content_uri || res,
+        size: buffer.length
+      };
+    };
+
+    const tag = autoTagger(senderId, this);
+
+    let res;
+    let localFilePath = '/tmp/matrix_bridge_tempfile';
```

Comment:
It may be wise to add a random string suffix to the filename, so that if multiple large file events are received at about the same time, and this runs concurrently for each, the two get separate files and there's no collision or corruption.

I don't know whether it does run concurrently now, but if it doesn't, it may in the future, and this could be a problem then.

Reply:
Ah, that's a good call and not something I thought about. Changes made.